src/java.base/share/native/launcher/main.c

```diff
@@ -93,6 +93,64 @@ WinMain(HINSTANCE inst, HINSTANCE previnst, LPSTR cmdline, int cmdshow)
     __initenv = _environ;
 
 #else /* JAVAW */
+
+#include <sys/wait.h>
+
+static int is_checkpoint = 0;
+
+static void parse_checkpoint(const char *arg) {
+    if (!is_checkpoint) {
+        const char *checkpoint_arg = "-XX:CRaCCheckpointTo";
+        const int len = strlen(checkpoint_arg);
+        if (0 == strncmp(arg, checkpoint_arg, len)) {
+            is_checkpoint = 1;
+        }
+    }
+}
+
+static pid_t g_child_pid = -1;
+
+static int wait_for_children() {
+    int status = 0;
+    pid_t pid;
+    do {
+        int st = 0;
+        pid = wait(&st);
+        if (pid == g_child_pid && WIFEXITED(st)) {
+            status = WEXITSTATUS(st);
```

Comment:
Sorry for nit-picking, but now if the java was killed (`WIFEXITED == false`) we won't update status and will return `0`, which does not look correct. `restorewait` in this situation returns `1`, although better, also does not look perfect. Here I suggest be at least consistent with restorewait.

Or we can fix restorewait as well, indicating being killed by returning `128+signal`, as described in the bash manual [1]. How does it sound?

> When a command terminates on a fatal signal N, bash uses the value of 128+N as the exit status.

[1] https://linux.die.net/man/1/bash

Reply:
To make things iterative, I suggest to implement wait_for_children() in the same way as restorewait() for now, and then create the next PR to make appropriate changes related to signal handling and returning exit codes.

Reply:
OK, makes sense for me as well.